### DISCUSSION TITLE
Use primaryKey() for update ids columns in ActiveRecord->insertInternal()

### DISCUSSION THREAD

**Discussion by nixonchern** [yiisoft/yii2#20245] [Ideas]
Hello! I have encountered a problem with obtaining an ID in my model by method `afterSave()`. After research problem i found in ActiveRecord->insertInternal():
```
        if (($primaryKeys = static::getDb()->schema->insert(static::tableName(), $values)) === false) {
            return false;
        }
        foreach ($primaryKeys as $name => $value) {
            $id = static::getTableSchema()->columns[$name]->phpTypecast($value);
            $this->setAttribute($name, $id);
            $values[$name] = $id;
        }
```
My table schema doesnt have PK. Because i use mysql VIEW, not TABLE. I know that not good use VIEW for model, but I have to do it.
So i believe "YII2" should be no restrictions for developers.

I suggest use code:

```
if (($primaryKeys = $this->primaryKey() ?: static::getDb()->schema->insert(static::tableName(), $values)) === false) {
    return false;
}
```

Sorry if I distract you from you work. I just want to bring something **good** to the community! Have a nice day! Maybe it use in yii3?

**Comment by bizley** [comment]
Why do you need to use afterSave() with a view?

**Reply by nixonchern** [reply]
I send by socket new information about model after save. 

I fix my problem:
In my case i just use another model which work with TABLE and now i have scheme.

**Reply by nixonchern** [reply]
I think if i can create some row in view by "sql" then i can do it by yii2. But I ran into a limitation

**Reply by nixonchern** [reply]
Same problem i found on this questions: https://stackoverflow.com/questions/34380688/yii-2-active-record-class-does-not-get-id-after-save

**Reply by nixonchern** [reply]
Its not so important problem for me. I fix it in my case. But it's unobvious behavior for me if i set $this->primaryKey() in my model. Thank you for your attention!

**Comment by rob006** [comment]
I don't think that this is related to lack of primary key. The problem is that `insert()` returns `false`, which means that query failed (it should return empty array for successful insert to table without PK). This may be related to the fact that you're inserting to view (make sure that related records are actually inserted) and caused by some quirk/bug in PDO/MySQL, but I don't think that Yii will be able do something about it.

You may try `Yii::$app->db->createCommand()->insert($table, $data)->execute()` and check what it returns.
